src/portfft/descriptor.hpp

```diff
@@ -231,16 +231,23 @@ class committed_descriptor {
   };
 
   struct dimension_struct {
-    std::vector<kernel_data_struct> kernels;
+    std::vector<kernel_data_struct> forward_kernels;
+    std::vector<kernel_data_struct> backward_kernels;
+    std::vector<kernel_data_struct> transpose_kernels;
     std::shared_ptr<IdxGlobal> factors_and_scan;
     detail::level level;
     std::size_t length;
     Idx used_sg_size;
     Idx num_batches_in_l2;
     Idx num_factors;
 
-    dimension_struct(std::vector<kernel_data_struct> kernels, detail::level level, std::size_t length, Idx used_sg_size)
-        : kernels(kernels), level(level), length(length), used_sg_size(used_sg_size) {}
+    dimension_struct(std::vector<kernel_data_struct> forward_kernels, std::vector<kernel_data_struct> backward_kernels,
+                     detail::level level, std::size_t length, Idx used_sg_size)
+        : forward_kernels(forward_kernels),
+          backward_kernels(backward_kernels),
```

Comment:
`std::move`?

Reply:
I would expect the compiler to treat them as temporary variables and treat them as rvalues, especially how we are using the `build_w_spec_const`, would it not ?

Although, I suppose we could change `dimensions.push_back` to `emplace_back` [here](https://github.com/codeplaysoftware/portFFT/blob/7dee1fbe640d889029e1a7381d0f0e8d0b596492/src/portfft/descriptor.hpp#L744)

Reply:
I think so - this [Stackoverflow question](https://stackoverflow.com/questions/23929800/is-stdmove-really-needed-on-initialization-list-of-constructor-for-heavy-membe) is similar I think.

I think in the second case, it might not be needed becuase the function returns an rvalue and the dimensions struct should already have rvalue assingment by default? I'm not totally sure.

Reply:
> it might not be needed becuase the function returns an rvalue and the dimensions struct should already have rvalue assingment by default?

That's what I am expecting, but using `emplace_back` would ensure move semantics are used. IIRC, `push_back` only uses copies

Reply:
(push_back)[https://en.cppreference.com/w/cpp/container/vector/push_back] can move.

Reply:
I think `emplace_back` is only really helpful when you are not using the copy constructor, so it probably won't help but it also won't hurt runtime performance. Might be slightly more complicated for the compiler though.

Reply:
using std::move now